nemo/core/neural_factory.py

```diff
@@ -36,10 +37,117 @@
+
+    for i in reversed(indices_to_remove):
+        hooks.pop(i)
+
+    _top_sorted_modules = []
+    all_nodes = {}
+
+    # extract all nodes to all_nodes set
+    hooks_lst = list(hooks)
+    while len(hooks_lst) > 0:
+        # take nmtensor from the end of the list
+        nmtensor = hooks_lst.pop()
+        producer_args = nmtensor.producer_args
+
+        node = create_node(nmtensor.producer, producer_args)
+        # Store nmtensor as an output of its producer
+        # first make sure all keys are present per output port
+        # and nm is inside all_nodes
+        if node not in all_nodes:
+            all_nodes[node] = {k: None for k in nmtensor.producer.output_ports}
+        # second, populate output port with current nmtensor
+        # where applicable
+        all_nodes[node][nmtensor.name] = nmtensor
+        processed_nmtensors.add(nmtensor)
+
+        new_tensors = set()
+        if producer_args is not None and producer_args != {}:
+            for _, new_nmtensor in producer_args.items():
+                if new_nmtensor not in processed_nmtensors:
+                    new_tensors.add(new_nmtensor)
+
+        # TODO
+        if cached_training_state:
+            for name, input_nmtensor in producer_args.items():
+                if cached_training_state.check_tensor_cached(input_nmtensor.unique_name):
+                    new_tensors.remove(input_nmtensor)
+
+        for new_nmtensor in new_tensors:
+            # put in the start of list
+            hooks_lst.insert(0, new_nmtensor)
+
+    all_node_with_output = []
+    # Iterate over all_nodes to create new nodes that include its output
+    # now all nodes have (module, input tensors, output tensors)
+    for node in all_nodes:
+        all_node_with_output.append(tuple((node[0], node[1], all_nodes[node])))
+
+    processed_nodes = []
+    while len(all_node_with_output) > 0:
+        for node in all_node_with_output.copy():
+            # if node's in_degree is zero it can be added to
+            # _top_sorted_modules
+            # this will also reduce in_degree of its children
+            if is_in_degree_zero(node, processed_nodes, cached_training_state):
+                _top_sorted_modules.append(node)
+                processed_nodes.append((node[0], node[1]))
+                all_node_with_output.remove(node)
+
+    # Create top_sorted_modules aka callchain
+    top_sorted_modules = []
+    for i, m in enumerate(_top_sorted_modules):
+        top_sorted_modules.append((m[0], dict(m[1]), m[2]))
+        # Ensure that there is only one dataset in callchain
+        if i > 0 and isinstance(m[0], DataLayerNM):
+            raise ValueError("There were more than one DataLayer NeuralModule inside your DAG.")
+
+        # TODO
+        if cached_training_state and isinstance(m[0], DataLayerNM):
+            raise ValueError("Could not compute tensor from current cached training state.")
+
+    return top_sorted_modules
```

Comment:
I really want to emphasize that this code is obsolete. Please note that neural graphs have this tiny property called steps.

For example, for graph:
```
with NeuralGraph(operation_mode=OperationMode.training) as g0:
    x, t = dl()
    p = m2(x=x)
    lss = loss(predictions=p, target=t)
```

self.steps contains:
```
    * Steps (3):
        0. realfunctiondatalayer0
        1. taylornet0
        2. mseloss0
```

Reply:
Again as actions is not based on NeuralGraphs yet, I see no reason why this would be helpful for me.

Reply:
Since, our call DAG is not generated from NeuralGraphs inside of actions, I cannot use the step property.

Reply:
I think you will generate exactly the same DAG, and end up with the same order... But let's take it out of the scope of this PR